src/ctc_network.c

```diff
@@ -0,0 +1,620 @@
+
+    int data_payload_size = 0;
+    int data_size = 0;
+    int retval;
+
+    if (IS_FAILURE (make_ctcp_header (op_id, op_param, control_session, job_session, header_data, &ctcp.header)))
+    {
+        goto error;
+    }
+
+    if (IS_NOT_NULL (data))
+    {
+        if (IS_FAILURE (make_ctcp_data_payload (ctcp.data_payload, data, header_data)))
+        {
+            goto error;
+        }
+
+        data_payload_size = header_data;
+    }
+
+    data_size = sizeof (CTCP_HEADER) + data_payload_size;
+
+    if (op_id != CTCP_CREATE_JOB_SESSION)
+    {
+        retval = write (control_session->sockfd, &ctcp, data_size);
+        if (retval == -1 || retval < data_size)
+        {
+            goto error;
+        }
+    }
+    else
+    {
+        retval = write (job_session->sockfd, &ctcp.header, data_size);
+        if (retval == -1 || retval < data_size)
+        {
+            goto error;
+        }
+    }
+
+    return CTC_SUCCESS;
+
+error:
+
+    return CTC_FAILURE;
+}
+
+int check_operation_id (CTCP_HEADER *ctcp_header, CTCP_OP_ID op_id)
+{
+    return ctcp_header->op_id == op_id ? CTC_SUCCESS : CTC_FAILURE;
+}
+
+int check_result_code (CTCP_HEADER *ctcp_header)
+{
+    if (ctcp_header->op_param_or_result_code != CTC_RC_SUCCESS &&
+        ctcp_header->op_param_or_result_code != CTC_RC_SUCCESS_FRAGMENTED)
+    {
+        goto error;
+    }
+
+    return CTC_SUCCESS;
+
+error:
+
+    return CTC_FAILURE;
+}
+
+int check_protocol_version (CTCP_HEADER *ctcp_header)
+{
+    if (ctcp_header->version[0] != CTCP_MAJOR_VERSION)
+    {
+        goto error;
+    }
+
+    return CTC_SUCCESS;
+
+error:
+
+    return CTC_FAILURE;
+}
+
+int check_job_desc (CTCP_HEADER *ctcp_header, JOB_SESSION *job_session)
+{
+    if (job_session != NULL && job_session->job_desc != -1)
+    {
+        if (ctcp_header->job_desc != job_session->job_desc)
+        {
+            goto error;
+        }
+    }
+
+    return CTC_SUCCESS;
+
+error:
+
+    return CTC_FAILURE;
+}
+
+int check_session_gid (CTCP_HEADER *ctcp_header, CONTROL_SESSION *control_session)
+{
+    if (control_session->session_gid != -1)
+    {
+        if (ctcp_header->session_gid != control_session->session_gid)
+        {
+            goto error;
+        }
+    }
+
+    return CTC_SUCCESS;
+
+error:
+
+    return CTC_FAILURE;
+}
+
+int set_session_gid (CONTROL_SESSION *control_session, int session_gid)
+{
+    if (session_gid < 0)
+    {
+        goto error;
+    }
+
+    control_session->session_gid = session_gid;
+
+    return CTC_SUCCESS;
+
+error:
+
+    return CTC_FAILURE;
+}
+
+void set_job_desc (JOB_SESSION *job_session, unsigned short job_desc)
+{
+    job_session->job_desc = job_desc;
+}
+
+int recv_ctcp_header (CTCP_OP_ID op_id, CONTROL_SESSION *control_session, JOB_SESSION *job_session, CTCP_HEADER *ctcp_header)
+{
+    int retval;
+
+    if (op_id != CTCP_CREATE_JOB_SESSION_RESULT)
+    {
+        retval = read (control_session->sockfd, ctcp_header, sizeof (CTCP_HEADER));
```

Comment:
Personally, I prefer wrapped netIO instead use of 'read' system call directly.
For example, there is a function 'net_read_stream ( )' at 'src/broker/cas_network.c' in CUBRID.
Because single 'read' may read insufficient number of bytes that a user expect to, 
and there may need another 'read' to get rest of data from network.
I think you can extract part of function code from CUBRID, it may not be the best but proven.

Reply:
I will consider this review.